cimr/processor/utils.py

```diff
@@ -106,6 +106,13 @@ def remove_palindromic(df):
         logging.error(f' alleles need to be indicated.')
         sys.exit(1)
 
+
+    # Check the genotype column to remove palindromic alleles.
+    # Palindromic alleles are pairs of complementary nucleotides.
+    # Strand flips cannot be distinguished and may bias results in
+    # downstream applications where accurate strand prediction is
+    # necessary. e.g. imputation or haplotype-based analysis.
+    # A and T or G and C pairs are removed from the data.
     df = df.loc[~((df.g == 'AT') | (df.g == 'TA') | (df.g == 'CG') | (df.g == 'GC'))]
```

Comment:
Thank you for the comments. I am not familiar with the data type of `df.loc`. My concern is that `~True` is -2, and `~False` is -1, is this really what you want?

Reply:
More for context on bitwise operators: https://wiki.python.org/moin/BitwiseOperators

Reply:
? @dongbohu not sure if I understand. .loc is just an index of the dataframe.

Reply:
? @dongbohu not sure if it matters here. .loc is just an index of the dataframe. Can you give me an example where boolean values actually would matter in this particular context?

Reply:
What @dongbohu is pointing out is that unless you intended to use bitwise operators, you may not be getting what you intend.

> (df.g == 'AT') | (df.g == 'TA') | (df.g == 'CG') | (df.g == 'GC')

Here' you're using the bitwise or. I'm guessing that `==` here returns `True` or `False`, so this should work though from my read it looks more appropriate to use the keyword `or` instead of the bitwise operator. If these are not booleans, it could return something different than you expect as described in that python wiki page.

The bit that looks more likely to be an error is the bitwise complement to the True/False that comes out of this. It looks like `df = df.loc[~(...)]` is going to either be `df = df.loc[-2]` or `df = df.loc[-1]`. Is this what you intend?

Reply:
I think this may have the rationale, but I'm not familiar enough with pandas to know if this is what you want: https://stackoverflow.com/a/45772078

Reply:
@cgreene yes, your reference also tells it. In pandas, element-wise operators are used for indexing. https://stackoverflow.com/questions/21415661/logical-operators-for-boolean-indexing-in-pandas

Reply:
Ok! I wasn't aware that pandas overloads bitwise operators as element-wise ones. Thanks!

Reply:
Me either. Thanks @ypar.